ocrd_utils/ocrd_utils/os.py

```diff
@@ -72,7 +73,12 @@ def get_ocrd_tool_json(executable, no_cache=False):
     """
     executable_name = Path(executable).name
     if no_cache or executable_name not in _ocrd_tool_cache:
-        ocrd_tool = loads(run([executable, '--dump-json'], stdout=PIPE).stdout)
+        try:
+            ocrd_tool = loads(run([executable, '--dump-json'], stdout=PIPE).stdout)
+        except FileNotFoundError:
+            ocrd_tool = {}
+        except JSONDecodeError:
+            ocrd_tool = {}
```

Comment:
FileNotFoundError: Seems to be mostly because I switch around venvs a lot and not properly deactivate (resetting $PATH) before switching, so this can go.

JSONDecodeError: True, this should explode for proper OCR-D processors, but for the discovery we get false positives like `ocrd-make` or `ocrd-import`.

Reply:
> JSONDecodeError: True, this should explode for proper OCR-D processors, but for the discovery we get false positives like `ocrd-make` or `ocrd-import`.

Isn't that the wrong place for this kind of distinction, though? If you came to `get_ocrd_tool_json` for an executable without even knowing whether it's an OCR-D CLI, you probably made a mistake earlier.

AFAICS the `cli.resmgr.list_available` context (with its `dynamic` option) is the only place where we introduce such ambiguity. That ends up in `OcrdResourceManager.list_available`, which expands a glob expression over all `ocrd-*` executables in `PATH`. IMO that would be the place to go directly at it (e.g. by catching `JSONDecodeError` and skipping the executable).

Reply:
The `FileNotFoundError` can also happen if you have resources for executables in `$HOME/.config/ocrd/resource_list.yml` :-/

E.g. in the `test_resource_manager_config_default`, we add a resource for non-existant processor `ocrd-foo` to the user resource list. When subsequently calling the resource_manager's `list_installed` method, we call `get_processor_resource_types`, which calls `get_ocrd_tool_json` for `ocrd-foo`, which fails.

Reply:
> When subsequently calling the resource_manager's `list_installed` method, we call `get_processor_resource_types`, which calls `get_ocrd_tool_json` for `ocrd-foo`, which fails.

Yes, but it only does that if `ocrd-foo` can be found on `PATH`:
https://github.com/OCR-D/core/blob/ef7981a1707122208b1b0523c92d9a7d5ca05efc/ocrd_utils/ocrd_utils/os.py#L162-L166

To me, again, it looks preferable to catch the exception by the caller, in this case `OcrdResourceManager.list_installed`. 

But (if properly documented) it can also stay in `get_ocrd_tool_json`.

Reply:
3f4d6c7b drops the catches in `get_ocrd_tool_json` and adds a try-except for FNFE  in `ocrd_utils.os.get_all_resources`.